### PR TITLE
Fix SIMD handle issues

### LINKED ISSUE #23159: Vector<T> regression, BadImageFormatException in release
```
C:\Work\BadImageFormat>dotnet run
True

C:\Work\BadImageFormat>dotnet run -c Release

Unhandled Exception: System.BadImageFormatException: An attempt was made to load a program with an incorrect format. (0x8007000B)
   at Program.BytesOrdinalEqualsStringAndAscii(String previousValue, Span`1 newValue)
   at Program.Main(String[] args) in C:\Work\BadImageFormat\Program.cs:line 15
```

```
System.BadImageFormatException
  HResult=0x8007000B
  Message=An attempt was made to load a program with an incorrect format. (0x8007000B)
  Source=BadImageFormat
  StackTrace:
   at Program.BytesOrdinalEqualsStringAndAscii(String previousValue, Span`1 newValue) in C:\Work\BadImageFormat\Program.cs:line 136
   at Program.Main(String[] args) in C:\Work\BadImageFormat\Program.cs:line 16
```
Repo
```csharp
using System;
using System.Diagnostics;
using System.Numerics;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Text;

class Program
{
    static void Main(string[] args)
    {
        var str = "application/json,text/html;q=0.9,application/xhtml+xml;q=0.9,application/xml;q=0.8,*/*;q=0.7";
        var span = Encoding.ASCII.GetBytes(str).AsSpan();

        Console.WriteLine(BytesOrdinalEqualsStringAndAscii(str, span));
    }

    [MethodImpl(MethodImplOptions.AggressiveOptimization)]
    public unsafe static bool BytesOrdinalEqualsStringAndAscii(string previousValue, Span<byte> newValue)
    {
        // We just widen the bytes to char for comparision, if either the string or the bytes are not ascii
        // this will result in non-equality, so we don't need to specifically test for non-ascii.
        Debug.Assert(previousValue.Length == newValue.Length);

        // Use IntPtr values rather than int, to avoid unnessary 32 -> 64 movs on 64-bit.
        // Unfortunately this means we also need to cast to byte* for comparisions as IntPtr doesn't
        // support operator comparisions (e.g. <=, >, etc).
        // Note: Pointer comparision is unsigned, so we use the compare pattern (offset + length <= count)
        // rather than (offset <= count - length) which we'd do with signed comparision to avoid overflow.
        var count = (IntPtr)newValue.Length;
        var offset = (IntPtr)0;

        ref var bytes = ref MemoryMarshal.GetReference(newValue);
        ref var str = ref MemoryMarshal.GetReference(previousValue.AsSpan());

        do
        {
            // If Vector not-accelerated or remaining less than vector size
            if (!Vector.IsHardwareAccelerated || (byte*)(offset + Vector<byte>.Count) > (byte*)count)
            {
                if (IntPtr.Size == 8) // Use Intrinsic switch for branch elimination
                {
                    // 64-bit: Loop longs by default
                    while ((byte*)(offset + sizeof(long)) <= (byte*)count)
                    {
                        if (Unsafe.Add(ref str, offset) != (char)Unsafe.Add(ref bytes, offset) ||
                            Unsafe.Add(ref str, offset + 1) != (char)Unsafe.Add(ref bytes, offset + 1) ||
                            Unsafe.Add(ref str, offset + 2) != (char)Unsafe.Add(ref bytes, offset + 2) ||
                            Unsafe.Add(ref str, offset + 3) != (char)Unsafe.Add(ref bytes, offset + 3) ||
                            Unsafe.Add(ref str, offset + 4) != (char)Unsafe.Add(ref bytes, offset + 4) ||
                            Unsafe.Add(ref str, offset + 5) != (char)Unsafe.Add(ref bytes, offset + 5) ||
                            Unsafe.Add(ref str, offset + 6) != (char)Unsafe.Add(ref bytes, offset + 6) ||
                            Unsafe.Add(ref str, offset + 7) != (char)Unsafe.Add(ref bytes, offset + 7))
                        {
                            goto NotEqual;
                        }

                        offset += sizeof(long);
                    }
                    if ((byte*)(offset + sizeof(int)) <= (byte*)count)
                    {
                        if (Unsafe.Add(ref str, offset) != (char)Unsafe.Add(ref bytes, offset) ||
                            Unsafe.Add(ref str, offset + 1) != (char)Unsafe.Add(ref bytes, offset + 1) ||
                            Unsafe.Add(ref str, offset + 2) != (char)Unsafe.Add(ref bytes, offset + 2) ||
                            Unsafe.Add(ref str, offset + 3) != (char)Unsafe.Add(ref bytes, offset + 3))
                        {
                            goto NotEqual;
                        }

                        offset += sizeof(int);
                    }
                }
                else
                {
                    // 32-bit: Loop ints by default
                    while ((byte*)(offset + sizeof(int)) <= (byte*)count)
                    {
                        if (Unsafe.Add(ref str, offset) != (char)Unsafe.Add(ref bytes, offset) ||
                            Unsafe.Add(ref str, offset + 1) != (char)Unsafe.Add(ref bytes, offset + 1) ||
                            Unsafe.Add(ref str, offset + 2) != (char)Unsafe.Add(ref bytes, offset + 2) ||
                            Unsafe.Add(ref str, offset + 3) != (char)Unsafe.Add(ref bytes, offset + 3))
                        {
                            goto NotEqual;
                        }

                        offset += sizeof(int);
                    }
                }
                if ((byte*)(offset + sizeof(short)) <= (byte*)count)
                {
                    if (Unsafe.Add(ref str, offset) != (char)Unsafe.Add(ref bytes, offset) ||
                        Unsafe.Add(ref str, offset + 1) != (char)Unsafe.Add(ref bytes, offset + 1))
                    {
                        goto NotEqual;
                    }

                    offset += sizeof(short);
                }
                if ((byte*)offset < (byte*)count)
                {
                    if (Unsafe.Add(ref str, offset) != (char)Unsafe.Add(ref bytes, offset))
                    {
                        goto NotEqual;
                    }
                }

                return true;
            }

            // do/while as entry condition already checked
            var AllTrue = new Vector<ushort>(ushort.MaxValue);
            do
            {
                var vector = Unsafe.As<byte, Vector<byte>>(ref Unsafe.Add(ref bytes, offset));
                Vector.Widen(vector, out var vector0, out var vector1);
                var compare0 = Unsafe.As<char, Vector<ushort>>(ref Unsafe.Add(ref str, offset));
                var compare1 = Unsafe.As<char, Vector<ushort>>(ref Unsafe.Add(ref str, offset + Vector<ushort>.Count));

                if (!AllTrue.Equals(
                    Vector.BitwiseAnd(
                        Vector.Equals(compare0, vector0),
                        Vector.Equals(compare1, vector1))))
                {
                    goto NotEqual;
                }

                offset += Vector<byte>.Count;
            } while ((byte*)(offset + Vector<byte>.Count) <= (byte*)count);

            // Vector path done, loop back to do non-Vector
            // If is a exact multiple of vector size, bail now
        } while ((byte*)offset < (byte*)count);

        return true;
    NotEqual:
        return false;
    }
}
```
Lastest nightly sdk from https://github.com/dotnet/core-sdk
```
C:\Work\BadImageFormat>dotnet --version
3.0.100-preview4-010681

C:\Work\BadImageFormat>dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100-preview4-010681
 Commit:    5561b45f06

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.18351
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\3.0.100-preview4-010681\

Host (useful for support):
  Version: 3.0.0-preview4-27509-04
  Commit:  1ca3ac91f6
```

### PR DESCRIPTION
**Pull request by CarolEidt**
PR #23052 was overly agressive at applying the provided class handle (which defaults to NO_CLASS_HANDLE). In addition, make a small change to `SIMDIntrinsicWiden`, as it doesn't need to go through the complexity of inferring a class handle when it already exists on the node it's duplicating.

Fix #23159

### CODE REVIEW

**Comment by fiigii** [comment]
This also fixes https://github.com/dotnet/coreclr/issues/23197, right?

**Comment by CarolEidt** [comment]
@dotnet-bot test Ubuntu x64 Checked CoreFX Tests
@dotnet-bot test Ubuntu x64 Formatting

**Comment by CarolEidt** [comment]
> This also fixes #23197, right?

Yes, I believe it does.

**Comment by CarolEidt** [comment]
@dotnet-bot test CentOS7.1 x64 Checked r2r_jitstress1

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib PTAL
cc @fiigii @tannergooding

File: src/jit/simd.cpp [hunk 1]
```diff
@@ -1055,20 +1055,32 @@ GenTree* Compiler::impSIMDPopStack(var_types type, bool expectAddr, CORINFO_CLAS
 
     bool isParam = false;
 
-    // If we have a ldobj of a SIMD local we need to transform it.
-    if (tree->OperGet() == GT_OBJ)
+    // If we are popping a struct type it must have a valid handle.
+    // If 'structHandle' is specifed:
```
Comment (mikedn) [inline src/jit/simd.cpp]: specified

File: src/jit/simd.cpp [hunk 2]
```diff
@@ -1055,20 +1055,32 @@ GenTree* Compiler::impSIMDPopStack(var_types type, bool expectAddr, CORINFO_CLAS
...
+    // - If we have an existing 'OBJ' we will change its handle if it doesn't match.
+    //   This can happen when we have a retyping of a vector that doesn't translate to any
+    //   actual IR.
+    // - (If it's not an OBJ and it's used in a parameter context where it is required,
+    //   impNormStructVal will add one).
+    // If 'structHandle' is not specified it is NO_CLASS_HANDLE), then we will use the
+    // handle on the stack value (see ti.GetClassHandleForValueClass() below).
+    //
+    if (structHandle != NO_CLASS_HANDLE)
```
Comment (mikedn) [inline src/jit/simd.cpp]: Most of the callers of `impSIMDPopStack` do not pass a struct handle. With this change `OBJ(ADDR(LCL_VAR))` won't be reduced to `LCL_VAR` anymore. Doesn't this cause CQ regressions?

Comment (CarolEidt) [reply src/jit/simd.cpp]: I'm not sure I understand the concern. This code will kick in only if the caller passed a valid class handle, in which case we want to ensure that the `OBJ`, if already present, will match. This addresses cases where we treat a vector conversion as a no-op, but don't want the `impNormStructType` code to assert when it encounters a non-matching handle.

File: src/jit/simd.cpp (hunk 1 above)
Comment (CarolEidt) [reply src/jit/simd.cpp]: Fixed

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib PTAL
(The CI tests all passed on the original commit; the latest commit changed only comments.)

File: src/jit/simd.cpp (hunk 2 above)
Comment (mikedn) [reply src/jit/simd.cpp]: > This code will kick in only if the caller passed a valid class handle

Yes, but most callers pass `NO_CLASS_HANDLE`. And before this change we had
```C++
        GenTree* addr = tree->gtOp.gtOp1;
        if ((addr->OperGet() == GT_ADDR) && isSIMDTypeLocal(addr->gtOp.gtOp1))
        {
            tree = addr->gtOp.gtOp1;
        }
```
removing the `OBJ` node. Now this code is only executing if a valid class handle is passed in.

Comment (CarolEidt) [reply src/jit/simd.cpp]: Ah, I see. I'll rework it.

**Comment by CarolEidt** [comment]
I ran diffs on the latest change (eliding the OBJ if no struct handle is specified), and there are zero diffs for x64/windows for benchmarks and frameworks.

**Comment by fiigii** [comment]
@dotnet-bot test CentOS7.1 x64 Checked r2r_jitstress1

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib ping

**Comment by CarolEidt** [comment]
@dotnet/dnceng - I've attempted to retry coreclr-ci, but it doesn't seem to trigger, and the remaining tests don't seem to complete (it's been 21 hours now)

**Pull request merged**